### ISSUE TITLE
Fatal error call to date() on null in woocommerce\src\Admin\API\Reports\Customers\DataStore.php:738

### ISSUE THREAD

**Issue by jantowell** [woothemes/woocommerce#35752]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

PHP Fatal error caused when date() is called in woocommerce\src\Admin\API\Reports\Customers\DataStore.php:738 when a user account with customer role has a user_registered value of null (000-00-00 00:00:00) the default value for this field in the wp_users table. This occurs when clicking "Update User" when editing said user account in WordPress

### Expected behavior

Expected the user account to be successfully saved.

### Actual behavior

User account fails to be updated and a critical error message is displayed with the fatal error and the message "There has been a critical error on this website. ...."

### Steps to reproduce

1. install clean wordpress site
2. install woocommerce and activate, skip store setup
3. Goto Users and Add New. Setup an account eg testcustomer and make the user role "customer", click "Add New User"
4. Goto database and edit the wp_users table for testcustomer to have user_registered 000-00-00 00:00:00
5. In WordPress Goto Users and Edit the testcustomer account. Change nothing and click "Update User" the fatal error will occur

Fatal error: Uncaught Error: Call to a member function date() on null in C:\xampp\htdocs\woocommercetest\wp-content\plugins\woocommerce\src\Admin\API\Reports\Customers\DataStore.php:738 Stack trace: #0 C:\xampp\htdocs\woocommercetest\wp-includes\class-wp-hook.php(308): Automattic\WooCommerce\Admin\API\Reports\Customers\DataStore::update_registered_customer(2) #1 C:\xampp\htdocs\woocommercetest\wp-includes\class-wp-hook.php(332): WP_Hook->apply_filters('', Array) #2 C:\xampp\htdocs\woocommercetest\wp-includes\plugin.php(517): WP_Hook->do_action(Array) #3 C:\xampp\htdocs\woocommercetest\wp-admin\user-edit.php(154): do_action('edit_user_profi...', 2) #4 {main} thrown in C:\xampp\htdocs\woocommercetest\wp-content\plugins\woocommerce\src\Admin\API\Reports\Customers\DataStore.php on line 738

6. Deactivate WooCommerce plugin
7. Edit testcustomer account and the error doesn't happen

### WordPress Environment

System report
`
### WordPress Environment ###

WordPress address (URL): https://localhost/woocommercetest
Site address (URL): https://localhost/woocommercetest
WC Version: 7.1.0
REST API Version: ✔ 7.1.0
WC Blocks Version: ✔ 8.7.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 768 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_AU
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.53 (Win64) OpenSSL/1.1.1n PHP/8.0.19
PHP Version: 8.0.19
PHP Post Max Size: 0 B
PHP Time Limit: 120
PHP Max Input Vars: 1000
cURL Version: 7.76.1
OpenSSL/1.1.1n

SUHOSIN Installed: –
MySQL Version: 10.4.24-MariaDB
Max Upload Size: 0 B
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.1.0
WC Database Prefix: wp_
Total Database Size: 3.84MB
Database Data Size: 2.40MB
Database Index Size: 1.44MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.50MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 7.1.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /?page_id=6
Cart: #7 - /?page_id=7
Checkout: #8 - /?page_id=8
My account: #9 - /?page_id=9
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty
Version: 1.5 (update to version 2.1 is available)
Author URL: https://en-au.wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build
personally we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-11-29 23:49:25 +00:00
Options: ✔
Notes: 35
Onboarding: skipped

### Action Scheduler ###

Complete: 2
Oldest: 2022-11-28 23:50:35 +0000
Newest: 2022-11-28 23:50:35 +0000

Pending: 2
Oldest: 2022-11-28 23:58:43 +0000
Newest: 2022-11-29 23:50:35 +0000

### Status report information ###

Generated at: 2022-11-28 23:57:52 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

In addition to all of the above the solution appears quite simple and is already used at line 573 in the same file ie
`$data['date_registered'] = $customer_user->get_date_created( 'edit' ) ? $customer_user->get_date_created( 'edit' )->date( TimeInterval::$sql_datetime_format ) : null;`

I am therefore suggesting the fix be to replace :
`'date_registered'  => $customer->get_date_created( 'edit' )->date( TimeInterval::$sql_datetime_format ),`
with
`'date_registered'  => $customer->get_date_created( 'edit' ) ? $customer->get_date_created( 'edit' )->date( TimeInterval::$sql_datetime_format ) : null,`

Thanks
Janette

**Comment by maketheboat** [comment]
we can replicate this bug on latest WP and latest WC and the suggested fix cleared up the issue.

**Comment by ericfri** [comment]
5752277-zen

**Comment by vladolaru** [comment]
5753895-zd-woothemes

**Comment by anastas10s-afk** [comment]
Can replicate the issue, using a test installation.
Follow-up on the thread here: 5704182-zd

**Comment by anastas10s-afk** [comment]
5703610-zen

**Comment by qgp-opic** [comment]
Hi,

I also encountered the same issue reported here https://wordpress.org/support/topic/woo-rest-api-customer-filter-by-all-roles-uncaught-error/#post-16243676

If the user_registered column of the users table has a value of 0000-00-00 00:00:00 it will throw an error when fetching customers using the REST API, ex: GET /wp-json/wc/v3/customers.

```
[21-Nov-2022 17:16:25 UTC] PHP Fatal error:  Uncaught Error: Call to a member function getTimestamp() on null in /home/customer/www/***/public_html/wp-content/plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-customers-controller.php:43
Stack trace:
```

This can be replicated on the latest version 7.1.0, but if we rollback to the previous version 7.0.0 everything is working fine.

Thanks,

Opic

**Comment by raifd** [comment]
5802542-zen

**Comment by kaushikasomaiya** [comment]
5855346-zen

**Comment by mlstolk** [comment]
Had the same issue, this fix resolves it properly and safely.

**Comment by kaushikasomaiya** [comment]
5962802-zen

**Comment by solstudioim** [comment]
Reported here as well on forum: https://wordpress.org/support/topic/php-fatal-error-uncaught-error-call-to-a-member-function-date-on-null/

**Comment by nicdwilson** [comment]
6005180-zen

**Comment by wyter** [comment]
6055556-zen

**Comment by larodrig** [comment]
I had the error a few months back and it fixed itself with a Woocommerce update. Now it was back.
I applied the fix and it worked like charm.
Thanks!

**Comment by tkintenn** [comment]
Same error listed above. WP 6.1.1 WC 7.5.1, everything up to date. 

replacing the array assignment for date_registered with a ternary if fixed the issue. Found on this thread: 
[https://wordpress.org/support/topic/php-fatal-error-uncaught-error-call-to-a-member-function-date-on-null/](https://wordpress.org/support/topic/php-fatal-error-uncaught-error-call-to-a-member-function-date-on-null/)

**Comment by Babylon1999** [comment]
One more in zd-6175862

**Comment by nicdwilson** [comment]
This error is now occurring during checkout in 6005180-zen. The site has an imported dataset of customers, some of which are set to the now invalid `0000-00-00 00:00:00` format.

The new order creation action during `process_checkout` is now triggering a customer update, and the error is seen as follows. I believe since there is now a risk of this crashing checkout, the priority of this item should be increased:

```
2023-04-18T02:45:12+00:00 CRITICAL Uncaught Error: Call to a member function date() on null in /srv/htdocs/wp-content/plugins/woocommerce/src/Admin/API/Reports/Customers/DataStore.php:750
Stack trace:
#0 /wordpress/core/6.2/wp-includes/class-wp-hook.php(310): Automattic\WooCommerce\Admin\API\Reports\Customers\DataStore::update_registered_customer(1392)
#1 /wordpress/core/6.2/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#2 /wordpress/core/6.2/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#3 /wordpress/core/6.2/wp-includes/user.php(2461): do_action('profile_update', 1392, Object(WP_User), Array)
#4 /wordpress/core/6.2/wp-includes/user.php(2582): wp_insert_user(Array)
#5 /srv/htdocs/wp-content/plugins/woocommerce/includes/data-stores/class-wc-customer-data-store.php(199): wp_update_user(Array)
#6 /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-data-store.php(196): WC_Customer_Data_Store->update(Object(WC_Customer))
#7 /srv/htdocs/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-data.php(226): WC_Data_Store->update(Object(WC_Customer))
#8 /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-checkout.php(1174): WC_Data->save()
#9 /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-checkout.php(1255): WC_Checkout->process_customer(Array)
#10 /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-ajax.php(485): WC_Checkout->process_checkout()
#11 /wordpress/core/6.2/wp-includes/class-wp-hook.php(308): WC_AJAX::checkout('')
#12 /wordpress/core/6.2/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters('', Array)
#13 /wordpress/core/6.2/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#14 /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-ajax.php(96): do_action('wc_ajax_checkou...')
#15 /wordpress/core/6.2/wp-includes/class-wp-hook.php(308): WC_AJAX::do_wc_ajax('')
#16 /wordpress/core/6.2/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(false, Array)
#17 /wordpress/core/6.2/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#18 /wordpress/core/6.2/wp-includes/template-loader.php(13): do_action('template_redire...')
#19 /wordpress/core/6.2/wp-blog-header.php(19): require_once('/wordpress/core...')
#20 /wordpress/core/6.2/index.php(17): require('/wordpress/core...')
#21 {main}
  thrown in /srv/htdocs/wp-content/plugins/woocommerce/src/Admin/API/Reports/Customers/DataStore.php on line 750
```

**Comment by thisissandip** [comment]
I am unable to replicate this WooCommerce 7.6.

I tried to change the values in the database and when I update the user, the user registration is updated to a random but correctly formatted date.

I tried with: https://wordpress.org/plugins/import-users-from-csv-with-meta/

And, I am unable to import users with the `000-00-00 00:00:00` registered date.

I do understand that if a plugin still allows `000-00-00 00:00:00` import,  this should be fixed and there should be a check if the user registered is null then it should be attached with a random date

**Comment by nicdwilson** [comment]
@thisissandip I would regard support for the date issue as WordPress technical debt. It is supported elsewhere is the same class. While the fatal is triggered by bad data, WooCommerce should handle this gracefully.

**Comment by kaushikasomaiya** [comment]
To replicate, the Strict mode/no_zero_in_date in SQL probably needs to be disabled temporarily: `SET GLOBAL sql_mode = '';`

Set 0 in a user_registered column that has a customer role:

![https://d.pr/i/oxiMDe](https://d.pr/i/oxiMDe+) 

Try editing the user account -> See the error.

**Comment by thisissandip** [comment]
Thanks, @kaushikasomaiya and @nicdwilson! Pinged the product team.

**Comment by xue28** [comment]
6217007-zen

**Comment by kathy330** [comment]
Got the same issue after updated to the latest woocommerce

**Comment by mattsherman** [comment]
I am able to reproduce this issue. Thanks for the helpful and detailed error reports @jantowell and all others on this thread. I have a PR up for review to address this issue in an upcoming version of WooCommerce.

https://github.com/woocommerce/woocommerce/pull/37907

**Comment by igorhereira** [comment]
6237255-zen

**Comment by marissaexplores** [comment]
[6257318-zen](6257318-zd-woothemes)

**Comment by happychait** [comment]
6272081-zen

**Comment by ChrissiePollock** [comment]
#6236950-zen

**Comment by mattsherman** [comment]
Re-closing this issue, as it is fixed in WooCommerce 7.7 (#37907).

If this problem still exists on a site running WooCommerce 7.7, we should re-open the issue.
